python-client/giskard/models/model_explanation.py

```diff
@@ -1,20 +1,180 @@
 import logging
 import warnings
-from typing import Callable, Dict, List, Any
+from typing import Callable, Dict, List, Any, Tuple, Iterable
 
+import wandb
 import numpy as np
 import pandas as pd
 
 from giskard.datasets.base import Dataset
-from giskard.ml_worker.utils.logging import timer
 from giskard.models.base import BaseModel
+from giskard.ml_worker.utils.logging import timer
 
 warnings.filterwarnings("ignore", message=".*The 'nopython' keyword.*")
 import shap  # noqa
 
 logger = logging.getLogger(__name__)
 
 
+def explain_full(model: BaseModel, dataset: Dataset, input_data: pd.DataFrame) -> Tuple[np.ndarray, shap.Explainer]:
+    """Perform SHAP values calculation for samples of a given dataset."""
+
+    def prepare_df(df: pd.DataFrame) -> pd.DataFrame:
+        """Prepare dataframe for an inference step."""
+        df = model.prepare_dataframe(df, column_dtypes=dataset.column_dtypes, target=dataset.target)
+
+        if dataset.target in df.columns:
+            prepared_dataset = Dataset(df, column_types=dataset.column_types, target=dataset.target)
+        else:
+            prepared_dataset = Dataset(df, column_types=dataset.column_types)
+
+        # Make sure column order is the same as in the dataset.df.
+        columns_original_order = (
+            model.meta.feature_names
+            if model.meta.feature_names
+            else [c for c in dataset.df.columns if c in prepared_dataset.df.columns]
+        )
+        prepared_df = prepared_dataset.df[columns_original_order]
+
+        return prepared_df
+
+    # Prepare background sample to be used in the KernelSHAP.
+    background_df = model.prepare_dataframe(dataset.df, dataset.column_dtypes, dataset.target)
+    background_sample = background_example(background_df, dataset.column_types)
+
+    # Prepare input data for explanation.
+    input_df = prepare_df(input_data)
+
+    # Obtain SHAP explanations.
+    explainer = shap.KernelExplainer(
+        model=model.predict_df, data=background_sample, feature_names=input_df.columns, keep_index=True
```

Comment:
I would rather pass `model=model(predict).raw` just for the case where if the predictions was cached, we don't need to calculate them again.

Reply:
I do not understand. 'model' param accepts a reference to a function, but 'model(predict).raw' is a result of a function execution, isn't it?

Reply:
Typo: model.predict(dataset).raw

Reply:
Sorry, in fact, what you wrote initially is correct!